helper/pluginutil/tls.go

```diff
@@ -0,0 +1,229 @@
+package pluginutil
+
+import (
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"crypto/tls"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"math/big"
+	mathrand "math/rand"
+	"net/url"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/SermoDigital/jose/jws"
+	"github.com/hashicorp/errwrap"
+	uuid "github.com/hashicorp/go-uuid"
+	"github.com/hashicorp/vault/api"
+)
+
+var (
+	// PluginUnwrapTokenEnv is the ENV name used to pass unwrap tokens to the
+	// plugin.
+	PluginUnwrapTokenEnv = "VAULT_UNWRAP_TOKEN"
+)
+
+// generateSignedCert is used internally to create certificates for the plugin
+// client and server. These certs are signed by the given CA Cert and Key.
+func GenerateCert() ([]byte, *ecdsa.PrivateKey, error) {
+	key, err := ecdsa.GenerateKey(elliptic.P521(), rand.Reader)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	host, err := uuid.GenerateUUID()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	template := &x509.Certificate{
+		Subject: pkix.Name{
+			CommonName: host,
+		},
+		DNSNames: []string{host},
+		ExtKeyUsage: []x509.ExtKeyUsage{
+			x509.ExtKeyUsageClientAuth,
+			x509.ExtKeyUsageServerAuth,
+		},
+		KeyUsage:     x509.KeyUsageDigitalSignature | x509.KeyUsageKeyEncipherment | x509.KeyUsageKeyAgreement,
+		SerialNumber: big.NewInt(mathrand.Int63()),
```

Comment:
Yes and no...in theory you should be able to identify any certificate uniquely via authority ID and serial, for whatever uses you may have. CRLs only take serials into account but that mostly means if you have duplicates you may block more than one rather than just one cert, so not a security hole really.

You are correct that if you restarted Vault you'd get the same sequence but the CA cert/key would be different, hence the authority ID would be different, so it doesn't really matter. No harm to randomize though!